geoh5py/shared/cut_by_extent.py

```diff
@@ -0,0 +1,551 @@
+    """
+    Create a coordinate system from a quad with the last vertex as origin.
+
+    Constructs a plane coordinate system (origin, u-vector, v-vector, normal)
+    from four vertices where the last vertex serves as the origin.
+
+    :param vertices: Array of shape (4, 3) containing the four vertices of the quad.
+
+    :return: Tuple containing (origin, u-vector, v-vector, normal vector).
+
+    :raises ValueError: If vertices is not a (4, 3) array.
+    """
+    if vertices.shape != (4, 3):
+        raise ValueError("vertices must be a (4, 3) array")
+
+    vertices = np.asarray(vertices)
+
+    p0 = vertices[3]
+    u = normalize(vertices[2] - p0)
+    v = normalize(vertices[0] - p0)
+    n = normalize(np.cross(u, v))
+    return p0, u, v, n
+
+
+def intersect_segment_plane(
+    a: np.ndarray,
+    b: np.ndarray,
+    p0: np.ndarray,
+    n: np.ndarray,
+    *,
+    eps_plane: float = 1e-5,
+) -> np.ndarray | None:
+    """
+    Compute intersection of a line segment with a plane.
+
+    Finds the intersection point between a line segment defined by points A and B
+    and a plane defined by point P0 and normal vector n.
+
+    :param a: Start point of the line segment.
+    :param b: End point of the line segment.
+    :param p0: Point on the plane.
+    :param n: Normal vector of the plane.
+    :param eps_plane: Tolerance for plane distance calculations.
+
+    :return: Intersection point if it exists within the segment, None otherwise.
+    """
+    da = np.dot(a - p0, n)
+    db = np.dot(b - p0, n)
+
+    # Both on same side or both almost on plane → treat as no crossing
+    if da * db > 0 or (abs(da) < eps_plane and abs(db) < eps_plane):
+        return None
+
+    denom = da - db
+    if abs(denom) < eps_plane:
+        return None
+
+    t = da / denom
+    if 0.0 <= t <= 1.0:
+        return a + t * (b - a)
+    return None
+
+
+def clean_extent_for_intersection(
+    extent: np.ndarray, vertices: np.ndarray
+) -> np.ndarray:
+    """
+    Clean and prepare extent array for 3D intersection calculations.
+
+    :param extent: Input extent array, shape (2, 2) or (2, 3)
+    :param vertices: Array of vertices to extract Z bounds from, shape (N, 3)
+
+    :return: Cleaned extent array with shape (2, 3)
+    """
+    if not isinstance(extent, np.ndarray):
+        raise TypeError("Expected a numpy array of extent values.")
+
+    if not extent.ndim == 2 or extent.shape not in [(2, 3), (2, 2)]:
+        raise TypeError("Expected a 2D numpy array with 2 points and 2 or 3 columns")
+
+    z_coords = vertices[:, 2]
+    zmin, zmax = float(z_coords.min()), float(z_coords.max())
+
+    if extent.shape[1] == 2:
+        # For 2D extents, add Z bounds with slight expansion for safety
+        z_bounds = np.array([[zmin - 1], [zmax + 1]], dtype=np.float64)
+        extent = np.column_stack([extent.astype(np.float64), z_bounds])
+
+    return extent
+
+
+#  pylint: disable=too-many-locals
+def compute_plane_box_intersections(
```

Comment:
Let's see if we can clean up the function to avoid too many locals
```suggestion
def plane_box_intersection(
```